### PR TITLE
Feature/add drmm

### PR DESCRIPTION
**Pull request by faneshion**

### CODE REVIEW

File: matchzoo/models/drmm_model.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import keras
+import keras.backend as K
+from keras.activations import softmax
+
+from matchzoo import engine
+
+logger = logging.getLogger(__name__)
+
+
+class DRMMModel(engine.BaseModel):
```
Comment (uduse) [inline matchzoo/models/drmm_model.py]: You can name the new models with out the "Model" suffix. i.e. use "DRMM" here.

File: matchzoo/models/drmm.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        :param input: The input tensor for multi-layer perceptron.
+        :return: The output tensor.
+        """
+        # shape = [B, L, H]
+        out = input
+        for idx, hidden_size in enumerate(self._params['hidden_sizes']):
+            out = keras.layers.Dense(hidden_size)(out)
+            out = keras.layers.Activation('tanh')(out)
+        # shape = [B, L, 1]
+        return out
```
Comment (bwanglzu) [inline matchzoo/models/drmm.py]: can we use `mlp` as a common layer?

File: matchzoo/models/drmm.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        """
+        Performs attention on the input.
+
+        :param attention_input: The input tensor for attention layer.
+        :param attention_mask: A tensor to mask the invalid values.
+        :return: The output tensor.
+        """
+        # shape = [B, L, 1]
+        dense_input = keras.layers.Dense(1, use_bias=False)(attention_input)
+        if attention_mask is not None:
```
Comment (bwanglzu) [inline matchzoo/models/drmm.py]:
```suggestion
        if attention_mask:
```

File: matchzoo/models/drmm.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+"""An implementation of DRMM Model."""
+import typing
+
+import keras
+import keras.backend as K
+from keras.activations import softmax
```
Comment (bwanglzu) [inline matchzoo/models/drmm.py]: use the same import level?

File: matchzoo/models/drmm.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+from matchzoo import engine
+
+
+class DRMM(engine.BaseModel):
```
Comment (bwanglzu) [inline matchzoo/models/drmm.py]: Use the same naming convension, e.g. `DSSMModel `, `CDSSMModel`.
```suggestion
class DRMMModel(engine.BaseModel):
```

Comment (faneshion) [reply matchzoo/models/drmm.py]: We will rename all models and remove the `Model`. In this way, all models can be called such as `matchzoo.models.DRMM`.

File: matchzoo/models/drmm.py (hunk 2 above)
Comment (yangliuy) [reply matchzoo/models/drmm.py]: Yes, I agree. MLP can be a common layer since it can be reused by other models.

File: matchzoo/models/drmm.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                        attention_mask: typing.Any = None):
+        """
+        Performs attention on the input.
+
+        :param attention_input: The input tensor for attention layer.
+        :param attention_mask: A tensor to mask the invalid values.
+        :return: The output tensor.
+        """
+        # shape = [B, L, 1]
+        dense_input = keras.layers.Dense(1, use_bias=False)(attention_input)
```
Comment (yangliuy) [inline matchzoo/models/drmm.py]: Is this dense layer for the dot product in the attention layer?

File: matchzoo/models/drmm.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        #   B = batch size (number of sequences)
+        #   D = embedding size
+        #   L = `input_left` sequence length
+        #   R = `input_right` sequence length
+        #   K = size of top-k
+
+        # Left input and right input.
+        # shape = [B, L]
+        # shape = [B, R]
+        query, doc = self._make_inputs()
```
Comment (yangliuy) [inline matchzoo/models/drmm.py]: Should the shape of doc be [B, dimension of histogram] ? The input should still be a pre-computed interaction matching matrix, correct ?

File: matchzoo/models/drmm.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        # shape = [B, 1, 1]
+        dot_score = keras.layers.Dot(axes=[1, 1])(
+            [attention_probs, dense_output])
+
+        flatten_score = keras.layers.Flatten()(dot_score)
+
+        x_out = self._make_output_layer()(flatten_score)
+        self._backend = keras.Model(inputs=[query, doc], outputs=x_out)
+
+    def attention_layer(self, attention_input: typing.Any,
```
Comment (uduse) [inline matchzoo/models/drmm.py]:
```suggestion
    def _attention_layer(self, attention_input: typing.Any,
```

File: matchzoo/models/drmm.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            # shape = [B, L, 1]
+            dense_input += adder
+        # shape = [B, L, 1]
+        attention_probs = keras.layers.Lambda(
+            lambda x: softmax(x, axis=1),
+            output_shape=self._params['input_shapes'][0]
+        )(dense_input)
+        return attention_probs
+
+    def multi_layer_perceptron(self, input: typing.Any):
```
Comment (uduse) [inline matchzoo/models/drmm.py]:
```suggestion
    def _multi_layer_perceptron(self, input: typing.Any):
```

File: matchzoo/models/drmm.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        embedding = self._make_embedding_layer()
+        # Process left input.
+        # shape = [B, L, D]
+        embed_query = embedding(query)
+        # shape = [B, L, 1]
+        attention_probs = self.attention_layer(embed_query)
+
+        # Process right input.
+        # shape = [B, L, 1]
+        dense_output = self.multi_layer_perceptron(doc)
```
Comment (uduse) [inline matchzoo/models/drmm.py]:
```suggestion
        dense_output = self._multi_layer_perceptron(doc)
```

File: matchzoo/models/drmm.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        # shape = [B, L]
+        # shape = [B, R]
+        query, doc = self._make_inputs()
+
+        embedding = self._make_embedding_layer()
+        # Process left input.
+        # shape = [B, L, D]
+        embed_query = embedding(query)
+        # shape = [B, L, 1]
+        attention_probs = self.attention_layer(embed_query)
```
Comment (uduse) [inline matchzoo/models/drmm.py]:
```suggestion
        attention_probs = self._attention_layer(embed_query)
```

**Pull request merged**